### ISSUE TITLE
Problem with FILTER on sparql endpoint

### ISSUE THREAD

**Issue by ofilangi** [comunica/comunica#304]
#### Issue type:

- :bug: Bug

____
#### Description:

Sparql query with a FILTER does not work with sparql endpoint (but seems work with LDF )

```
comunica-sparql http://fragments.dbpedia.org/2015-10/en 'PREFIX owl: <http://www.w3.org/2002/07/owl#> select * where { { ?c a owl:Class . }  FILTER (  ! BOUND(?c) || ! ( REGEX(str(?c), "^toto","i"))) } limit 10'

comunica-sparql http://fragments.dbpedia.org/2015-10/en 'PREFIX owl: <http://www.w3.org/2002/07/owl#> select * where { { ?c a owl:Class . }  FILTER (  ! (BOUND(?c) && ( REGEX(str(?c), "^toto","i")))) } limit 10'
```
The filters on theses two queries are equivalents and gives the same results (so works with LDF)

```
comunica-sparql http://data.bnf.fr/sparql 'PREFIX owl: <http://www.w3.org/2002/07/owl#> select * where { { ?c a owl:Class . }  FILTER (  ! BOUND(?c) || ! ( REGEX(str(?c), "^toto","i"))) } limit 10'

comunica-sparql http://data.bnf.fr/sparql 'PREFIX owl: <http://www.w3.org/2002/07/owl#> select * where { { ?c a owl:Class . }  FILTER (  ! (BOUND(?c) && ( REGEX(str(?c), "^toto","i")))) } limit 10'
```
The filters on theses two queries are equivalents and don't give the same results (so does not work with SPARQL endpoint). The second query give an empty result set and should give ten results.

____
#### Environment:

| software            | version
| ------------------- | -------
| Comunica Init Actor | 1.4.1
| node                | v10.4.1
| npm                 | 6.1.0
| yarn                | _Yarn is unavailable_
| Operating System    | linux (Linux 4.4.0-17134-Microsoft)

**Comment by rubensworks** [comment]
When comunica queries a single SPARQL endpoint (your last example), then the query will be sent as-is to the endpoint.
So it looks like http://data.bnf.fr/sparql (Virtuoso) might be producing incorrect results here. So the best solution would be to fix it there.

There is an alternative option to _not_ send the query as-is to the endpoint, and instead perform the FILTER client-side. This will be slower and will involve creating a custom comunica config for querying. If you're interested in this, please let me know here, and I'll give you some more details.

**Comment by ofilangi** [comment]
Dear Ruben,
if I use a curl command, it's work
`curl -H "Accept: application/json" http://data.bnf.fr/sparql  --data-urlencode query='PREFIX owl: <http://www.w3.org/2002/07/owl#> select * where { { ?c a owl:Class . }  FILTER (  ! (BOUND(?c) && ( REGEX(str(?c), "^toto","i")))) } limit 10'`
give
```
{ "head": { "link": [], "vars": ["c"] },
  "results": { "distinct": false, "ordered": true, "bindings": [
    { "c": { "type": "uri", "value": "http://www.w3.org/2002/07/owl#Thing" }},
    { "c": { "type": "uri", "value": "http://www.w3.org/2002/07/owl#Thing" }},
    { "c": { "type": "uri", "value": "http://www.w3.org/2002/07/owl#Nothing" }},
    { "c": { "type": "uri", "value": "http://www.w3.org/2002/07/owl#Nothing" }},
    { "c": { "type": "uri", "value": "http://www.w3.org/2000/01/rdf-schema#Class" }},
    { "c": { "type": "uri", "value": "http://www.w3.org/2000/01/rdf-schema#Class" }},
    { "c": { "type": "uri", "value": "http://www.w3.org/2000/01/rdf-schema#Class" }},
    { "c": { "type": "uri", "value": "http://www.w3.org/2000/01/rdf-schema#Class" }},
    { "c": { "type": "uri", "value": "http://www.w3.org/2000/01/rdf-schema#Class" }},
    { "c": { "type": "uri", "value": "http://www.w3.org/2000/01/rdf-schema#Class" }} ] } }
```
versus

`comunica-sparql http://data.bnf.fr/sparql 'PREFIX owl: <http://www.w3.org/2002/07/owl#> select * where { { ?c a owl:Class . }  FILTER (  ! (BOUND(?c) && ( REGEX(str(?c), "^toto","i")))) } limit 10'`
gives :
`[]`

**Comment by rubensworks** [comment]
Oh, that's interesting. So it's a bug in comunica after all!

I'll look into it soon.

**Comment by rubensworks** [comment]
Ok, so it looks like comunica internally transforms

```
FILTER ( ! (BOUND(?c) && ( REGEX(str(?c), "^toto","i"))))
```

to 

```
FILTER( ! (BOUND(?c)) && ( REGEX(STR(?c), "^toto", "i")))
```

which is incorrect, as it breaks the negation scope.

This seems to be a problem in @joachimvh's SPARQL algebra, reported there (#30).

**Comment by rubensworks** [comment]
This has been fixed a while ago in #30.
